karapace/kafka_admin.py

```diff
@@ -0,0 +1,290 @@
+    sasl_mechanism: str | None
+    sasl_plain_password: str | None
+    sasl_plain_username: str | None
+    security_protocol: str | None
+    ssl_cafile: str | None
+    ssl_certfile: str | None
+    ssl_keyfile: str | None
+    sasl_oauth_token_provider: TokenWithExpiryProvider
+
+
+class KafkaAdminClient(AdminClient):
+    def __init__(self, bootstrap_servers: Iterable[str] | str, **params: Unpack[AdminClientParams]) -> None:
+        super().__init__(self._get_config_from_params(bootstrap_servers, **params))
+        self._activate_callbacks()
+        self._verify_connection()
+
+    @staticmethod
+    def _get_config_from_params(bootstrap_servers: Iterable[str] | str, **params: Unpack[AdminClientParams]) -> dict:
+        if not isinstance(bootstrap_servers, str):
+            bootstrap_servers = ",".join(bootstrap_servers)
+
+        config: dict[str, int | str | Callable | None] = {
+            "bootstrap.servers": bootstrap_servers,
+            "client.id": params.get("client_id"),
+            "connections.max.idle.ms": params.get("connections_max_idle_ms"),
+            "metadata.max.age.ms": params.get("metadata_max_age_ms"),
+            "sasl.mechanism": params.get("sasl_mechanism"),
+            "sasl.password": params.get("sasl_plain_password"),
+            "sasl.username": params.get("sasl_plain_username"),
+            "security.protocol": params.get("security_protocol"),
+            "ssl.ca.location": params.get("ssl_cafile"),
+            "ssl.certificate.location": params.get("ssl_certfile"),
+            "ssl.key.location": params.get("ssl_keyfile"),
+        }
+        config = {key: value for key, value in config.items() if value is not None}
+
+        if "sasl_oauth_token_provider" in params:
+            config["oauth_cb"] = params["sasl_oauth_token_provider"].token_with_expiry
+
+        return config
+
+    def _activate_callbacks(self) -> None:
+        # Any client in the Confluent Kafka Python library needs poll called to
+        # trigger any callbacks registered (eg. for errors, OAuth tokens, etc.)
+        self.poll(timeout=0.0)
+
+    def _verify_connection(self) -> None:
+        """Attempts to call `AdminClient.list_topics` a few times.
+
+        The `list_topics` method is the only meaningful synchronous method of
+        the `AdminClient` class that can be used to verify that a connection and
+        authentication has been established with a Kafka cluster.
+
+        Just instantiating and initializing the admin client doesn't result in
+        anything in its main thread in case of errors, only error logs from another
+        thread otherwise.
+        """
+        for _ in range(3):
+            try:
+                self.list_topics(timeout=1)
+            except KafkaException as exc:
+                LOG.info("Could not establish connection due to: %s", exc)
+                continue
+            else:
+                break
+        else:
+            raise AuthenticationFailedError()
```

Comment:
issue: This looks devious. There are other things than authentication that could be the cause of inability to connect, aren't we hiding that here?

Reply:
I massaged this to be more or less the same as the old admin client when it comes to the exceptions raised. I'm not sure if we can get more detail out of the `confluent_kafka.admin.AdminClient`; even the error callback doesn't get much detailed information.